Data race in TestPeerConnection_Renegotation_RemoveTrack

```
=== RUN   TestPeerConnection_Renegotation_RemoveTrack
==================
WARNING: DATA RACE
Read at 0x00c00020e068 by goroutine 81:
  github.com/pion/webrtc/v2.(*SCTPTransport).generateDataChannelID.func1()
      /home/travis/gopath/src/github.com/pion/webrtc/sctptransport.go:345 +0xfa
  github.com/pion/webrtc/v2.(*SCTPTransport).generateDataChannelID()
      /home/travis/gopath/src/github.com/pion/webrtc/sctptransport.go:362 +0x1b6
  github.com/pion/webrtc/v2.(*DataChannel).open()
      /home/travis/gopath/src/github.com/pion/webrtc/datachannel.go:155 +0x672
  github.com/pion/webrtc/v2.(*PeerConnection).CreateDataChannel()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1298 +0x6fb
  github.com/pion/webrtc/v2.signalPair()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_test.go:47 +0xc1
  github.com/pion/webrtc/v2.TestPeerConnection_Renegotation_RemoveTrack()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:124 +0x5e8
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb

Previous write at 0x00c00020e068 by goroutine 48:
  github.com/pion/webrtc/v2.(*DataChannel).open()
      /home/travis/gopath/src/github.com/pion/webrtc/datachannel.go:160 +0x6e4
  github.com/pion/webrtc/v2.(*PeerConnection).startSCTP()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:905 +0x287
  github.com/pion/webrtc/v2.(*PeerConnection).startTransports.func1()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1576 +0x6a5

Goroutine 81 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1043 +0x660
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1285 +0xa6
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb
  testing.runTests()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1283 +0x527
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1200 +0x2ff
  main.main()
      _testmain.go:580 +0x347

Goroutine 48 (running) created at:
  github.com/pion/webrtc/v2.(*PeerConnection).startTransports()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1544 +0x184
  github.com/pion/webrtc/v2.(*PeerConnection).SetRemoteDescription()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:773 +0x9cc
  github.com/pion/webrtc/v2.signalPair()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_test.go:79 +0x3e8
  github.com/pion/webrtc/v2.TestPeerConnection_Renegotation_RemoveTrack()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:120 +0x51b
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb
==================
==================
WARNING: DATA RACE
Read at 0x00c00030ed58 by goroutine 81:
  github.com/pion/webrtc/v2.(*SCTPTransport).generateDataChannelID.func1()
      /home/travis/gopath/src/github.com/pion/webrtc/sctptransport.go:345 +0x163
  github.com/pion/webrtc/v2.(*SCTPTransport).generateDataChannelID()
      /home/travis/gopath/src/github.com/pion/webrtc/sctptransport.go:362 +0x1b6
  github.com/pion/webrtc/v2.(*DataChannel).open()
      /home/travis/gopath/src/github.com/pion/webrtc/datachannel.go:155 +0x672
  github.com/pion/webrtc/v2.(*PeerConnection).CreateDataChannel()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1298 +0x6fb
  github.com/pion/webrtc/v2.signalPair()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_test.go:47 +0xc1
  github.com/pion/webrtc/v2.TestPeerConnection_Renegotation_RemoveTrack()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:124 +0x5e8
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb

Previous write at 0x00c00030ed58 by goroutine 48:
  github.com/pion/webrtc/v2.(*DataChannel).open()
      /home/travis/gopath/src/github.com/pion/webrtc/datachannel.go:155 +0x6a1
  github.com/pion/webrtc/v2.(*PeerConnection).startSCTP()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:905 +0x287
  github.com/pion/webrtc/v2.(*PeerConnection).startTransports.func1()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1576 +0x6a5

Goroutine 81 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1043 +0x660
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1285 +0xa6
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb
  testing.runTests()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1283 +0x527
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:1200 +0x2ff
  main.main()
      _testmain.go:580 +0x347

Goroutine 48 (running) created at:
  github.com/pion/webrtc/v2.(*PeerConnection).startTransports()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:1544 +0x184
  github.com/pion/webrtc/v2.(*PeerConnection).SetRemoteDescription()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection.go:773 +0x9cc
  github.com/pion/webrtc/v2.signalPair()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_test.go:79 +0x3e8
  github.com/pion/webrtc/v2.TestPeerConnection_Renegotation_RemoveTrack()
      /home/travis/gopath/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:120 +0x51b
  testing.tRunner()
      /home/travis/.gimme/versions/go1.14.linux.amd64/src/testing/testing.go:992 +0x1eb
==================
    TestPeerConnection_Renegotation_RemoveTrack: testing.go:906: race detected during execution of test
--- FAIL: TestPeerConnection_Renegotation_RemoveTrack (0.44s)
```
Go: 1.14 https://travis-ci.org/github/pion/webrtc/jobs/667158306

--- a/datachannel.go
+++ b/datachannel.go
@@ -152,12 +152,10 @@ func (d *DataChannel) open(sctpTransport *SCTPTransport) error {
 	}
 
 	if d.id == nil {
-		generatedID, err := d.sctpTransport.generateDataChannelID(d.sctpTransport.dtlsTransport.role())
+		err := d.sctpTransport.generateAndSetDataChannelID(d.sctpTransport.dtlsTransport.role(), &d.id)
 		if err != nil {
 			return err
 		}
-
-		d.id = &generatedID
 	}
 
 	dc, err := datachannel.Dial(d.sctpTransport.association, *d.id, cfg)

--- a/sctptransport.go
+++ b/sctptransport.go
@@ -339,7 +339,7 @@ func (r *SCTPTransport) collectStats(collector *statsReportCollector) {
 	collector.Collect(stats.ID, stats)
 }
 
-func (r *SCTPTransport) generateDataChannelID(dtlsRole DTLSRole) (uint16, error) {
+func (r *SCTPTransport) generateAndSetDataChannelID(dtlsRole DTLSRole, idOut **uint16) error {
 	isChannelWithID := func(id uint16) bool {
 		for _, d := range r.dataChannels {
 			if d.id != nil && *d.id == id {
@@ -356,14 +356,15 @@ func (r *SCTPTransport) generateDataChannelID(dtlsRole DTLSRole) (uint16, error)
 
 	max := r.MaxChannels()
 
-	r.lock.RLock()
-	defer r.lock.RUnlock()
+	r.lock.Lock()
+	defer r.lock.Unlock()
 	for ; id < max-1; id += 2 {
 		if isChannelWithID(id) {
 			continue
 		}
-		return id, nil
+		*idOut = &id
+		return nil
 	}
 
-	return 0, &rtcerr.OperationError{Err: ErrMaxDataChannelID}
+	return &rtcerr.OperationError{Err: ErrMaxDataChannelID}
 }

--- a/sctptransport_test.go
+++ b/sctptransport_test.go
@@ -33,13 +33,14 @@ func TestGenerateDataChannelID(t *testing.T) {
 		{DTLSRoleServer, sctpTransportWithChannels([]uint16{1, 5}), 3},
 	}
 	for _, testCase := range testCases {
-		id, err := testCase.s.generateDataChannelID(testCase.role)
+		idPtr := new(uint16)
+		err := testCase.s.generateAndSetDataChannelID(testCase.role, &idPtr)
 		if err != nil {
 			t.Errorf("failed to generate id: %v", err)
 			return
 		}
-		if id != testCase.result {
-			t.Errorf("Wrong id: %d expected %d", id, testCase.result)
+		if *idPtr != testCase.result {
+			t.Errorf("Wrong id: %d expected %d", *idPtr, testCase.result)
 		}
 	}
 }

